src/librustc_mir/transform/const_prop.rs

```diff
@@ -364,8 +364,15 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
             Rvalue::Use(ref op) => {
                 self.eval_operand(op, source_info)
             },
+            Rvalue::Ref(_, _, ref place) => {
+                let src = self.eval_place(place, source_info)?;
+                self.use_ecx(source_info, |this| {
+                    let dest = this.ecx.allocate(place_layout, MemoryKind::Stack);
```

Comment:
I'm not *too* happy about this, but see no good way around it for now. Please add a `FIXME` comment since we'll be able to clean it up once we create our own `Machine` just for const prop.

Reply:
Oh, this can only happen for `promoted` right now, right? So, you could just do this for `Operand::Indirect` for now, which doesn't require an additional allocation.

Reply:
Which part are you not happy with? The allocation?

Reply:
yes

Reply:
What I mean is that you can return `None` in case you get a `Operand::Immediate` and `Some(Operand::Immediate(mplace.ptr.into()))` in case you have an `Operand::Indirect(mplace)`

Reply:
So, I tried this (which I think is the equivalent to the pseudo code you provided):

```rust
let src = self.eval_place(place, source_info)?;
let mplace = src.try_as_mplace().ok()?;
Some(ImmTy::from_scalar(mplace.ptr.into(), mplace.layout).into())
```

and it ICEs compiling libcore:

```
thread 'rustc' panicked at 'value should already be a valid const: EvalError { kind: type validation failed: encountered a pointer, but expected a valid enum discriminant, backtrace: None }', src/libcore/result.rs:999:5
```

Reply:
I can debug a bit more this evening or I can just put the FIXME in if you don't think it's critical.

Reply:
I worry that we'll end up with temporary allocations in the promoted `mir::Body`. These allocations will be gone after const prop finishes.

> and it ICEs compiling libcore:

the `mplace.layout` is now wrong. A pointer has pointer layout, but the `MPlaceTy`'s `layout` field is the `layout` of the pointee. `place_layout` is the right layout to use here I believe.